dev/com.ibm.ws.security.jca/src/com/ibm/ws/security/jca/internal/AuthDataServiceImpl.java

```diff
@@ -146,10 +155,52 @@ private AuthData getAuthData(String authDataAlias) throws LoginException {
         return authDataProviderRef.getService().getAuthData(authDataAlias);
     }
 
-    private Subject obtainSubject(ManagedConnectionFactory managedConnectionFactory, AuthData authData) {
-        Subject subject = createSubject(managedConnectionFactory, authData);
-        addInvocationSubjectPrincipal(subject);
-        optimize(subject);
+    private Subject obtainSubject(ManagedConnectionFactory managedConnectionFactory, AuthData authData) throws LoginException {
+        if (authData.getKrb5Principal() != null) {
+            return doKerberosLogin(authData.getKrb5Principal(), authData.getKrb5Keytab());
+        } else {
+            Subject subject = createSubject(managedConnectionFactory, authData);
+            addInvocationSubjectPrincipal(subject);
+            optimize(subject);
+            return subject;
+        }
+    }
+
+    private Subject doKerberosLogin(String principal, Path keytab) throws LoginException {
+        Subject subject = new Subject();
+        Krb5LoginModuleWrapper krb5 = new Krb5LoginModuleWrapper();
+        Map<String, String> options = new HashMap<String, String>();
+        Map<String, Object> sharedState = new HashMap<String, Object>();
+
+        options.put("isInitiator", "true");
+        options.put("refreshKrb5Config", "true");
+        options.put("doNotPrompt", "true");
+        // TODO: Do we want to store the key?
+        //options.put("storeKey", "true");
+        options.put("useKeyTab", "true");
+        // If no keytab path specified, still set useKeyTab=true because then the
+        // default JDK or default OS locations will be checked
+        if (keytab != null) {
+            options.put("keyTab", keytab.toAbsolutePath().toString());
+        }
+        options.put("principal", principal);
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            options.put("debug", "true");
+            Tr.debug(tc, "All kerberos config properties are: " + options);
+        }
+
+        krb5.initialize(subject, null, sharedState, options);
+        krb5.login();
+        krb5.commit();
+
+        // If the created Subject does not have a GSSCredential, then create one and
+        // associate it with the Subject
```

Comment:
We should call SubjectHelpet.getGSSCredentialFromSubject() method

Reply:
@utle the goal of this code block is to ensure that a GSSCredential is associated with the Subject after the method exits, because the JDBC code expects to find a GSSCredential on the subject later on. To accomplish this, first I needed to check if there was already a GSSCred on the subject, and if it's not already there, then we do call `SubjectHelper.getGSSCredentialFromSubject(subject);` on line 200 and add that GSSCredential to the subject.